### PR TITLE
JAV-495 Set default URL pattern

### PR DESCRIPTION
**Pull request by yhs0092**
see details in [https://servicecomb.atlassian.net/browse/JAV-495](https://servicecomb.atlassian.net/browse/JAV-495)

- Set default URL pattern when RestServletInitializer is invoked and URL pattern is not specified.
- Add a demo to demonstrate how a Spring Boot project is converted to a JavaChassis project.

### CODE REVIEW

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/src/main/java/io/servicecomb/demo/pojo/client/PojoClientTest.java [hunk 1]
```diff
@@ -105,7 +110,7 @@ private static void testEmpty(Test test) {
 
   private static void testNull(Test test) {
     TestMgr.check("code is 'null'", test.getTestString(null));
-    TestMgr.check(null, test.wrapParam(null));
+//    TestMgr.check(null, test.wrapParam(null));
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/src/main/java/io/servicecomb/demo/pojo/client/PojoClientTest.java]: lost this test case.

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/HelloImpl.java [hunk 2]
```diff
@@ -16,15 +16,28 @@
...
+import javax.ws.rs.core.MediaType;
+
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
 import io.servicecomb.demo.helloworld.greeter.Hello;
+import io.servicecomb.provider.pojo.RpcSchema;
 
+@RpcSchema(schemaId = "helloworld.Greeter")
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/HelloImpl.java]: why old one no need this annotations?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java [hunk 3]
```diff
@@ -19,24 +19,39 @@
...
+import org.springframework.web.bind.annotation.ResponseBody;
+
 import io.servicecomb.demo.server.Test;
 import io.servicecomb.demo.server.TestRequest;
 import io.servicecomb.demo.server.User;
 import io.servicecomb.provider.pojo.RpcSchema;
 import io.servicecomb.swagger.invocation.exception.InvocationException;
 
 @RpcSchema(schemaId = "server")
+@RequestMapping(path = "/pojo/rest", produces = MediaType.APPLICATION_JSON)
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java]: why RpcSchema need this annotations?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/handler/MyHandler.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(MyHandler.class);
+
+  public static final String SPLITPARAM_RESPONSE_USER_SUFFIX = "(modified by MyHandler)";
+
+  @Override
+  public void handle(Invocation invocation, AsyncResponse asyncResp) throws Exception {
+    LOGGER.info("If you see this log, that means this demo project has been converted to ServiceComb framework.");
+
+    invocation.next(response -> {
+      if (invocation.getOperationName().equals("splitParam")) {
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/handler/MyHandler.java]: schema+operation is a unique key
OperationMeta have some qualifiedName method, you can call one of them.

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/resources/microservice.yaml [hunk 5]
```diff
@@ -8,3 +8,11 @@ cse:
       address: http://127.0.0.1:30100
   rest:
     address: 0.0.0.0:8080
+  handler:
+    chain:
+      Provider:
+        default: myhandler
+servicecomb:
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/resources/microservice.yaml]: if you want to test auto urlPattern, should this be deleted?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/handler/MyHandler.java (hunk 4 above)
Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/handler/MyHandler.java]:
```java
    invocation.next(response -> {
      if (invocation.getOperationMeta().getSchemaQualifiedName().equals("server.splitParam")) {
```
the judgement logic has been changed like above. Is this what you mean?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/resources/microservice.yaml (hunk 5 above)
Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/resources/microservice.yaml]:
```yaml
servicecomb:
  rest:
    servlet:
      urlPattern: /*
```
I guess you mean the configuration above should be deleted, right?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/HelloImpl.java (hunk 2 above)
Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/HelloImpl.java]: In old edition, the schema of HelloImpl is configured in a xml file. But the schema of TestImpl is configured by annotation. After discussion with Liu Bao, we think that two kind of configuration existing in one demo is confusing, so I transform it into annotation config.

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java (hunk 3 above)
Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java]: I keep this annotation in code for the purpose that users can transform this demo back into a Spring Boot demo more conveniently.(And they can compare this two kind of framework)
But actually this annotation is redundant, I will remove it and create another PR.

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/src/main/java/io/servicecomb/demo/pojo/client/PojoClientTest.java (hunk 1 above)
Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/src/main/java/io/servicecomb/demo/pojo/client/PojoClientTest.java]: I commented out this line for debugging... I will retrieve it.

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java [hunk 6]
```diff
@@ -19,6 +19,14 @@
 import java.util.Arrays;
 import java.util.List;
 
+import javax.ws.rs.core.MediaType;
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-server/src/main/java/io/servicecomb/demo/pojo/server/TestImpl.java]: compile warning?

File: demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/pom.xml [hunk 7]
```diff
@@ -42,6 +46,11 @@
 			<groupId>org.springframework.cloud</groupId>
 			<artifactId>spring-cloud-netflix-core</artifactId>
 		</dependency>
+		<dependency>
```
Comment (wujimin) [inline demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/pom.xml]: client depend on server
then service's microservice.yaml in classpath too, which one will override another one, it's random, except you point the order.
even you point the order, it's bad practice.

all server schema deployed to client too. do you want this?

Comment (yhs0092) [reply demo/demo-spring-boot-transport/demo-spring-boot-pojo-client/pom.xml]: I have fixed this problem by removing the dependency in client side, and organized import of RestServletInitializer.java. Please review this commit.

**Comment by yhs0092** [comment]
@wujimin I close this pull request and reopen it, and the travis-ci build passed. I can't find out the failure cause.

**Pull request merged**